subgraphs/abracadabra/src/positions.ts

```diff
@@ -388,3 +202,234 @@ function getAccountBalance(
 
   return tryBalance.reverted ? BIGINT_ZERO : tryBalance.value;
 }
+
+export function getLiquidatePosition(
+  side: string,
+  marketId: string,
+  accountId: string
+): string {
+  const positionCounter = PositionCounter.load(
+    accountId.concat("-").concat(marketId).concat("-").concat(side)
+  );
+  if (!positionCounter) {
+    log.warning("No liquidation position found for account {} on market {}", [
+      accountId,
+      marketId,
+    ]);
+    return "";
+  }
+
+  const position = Position.load(
+    positionCounter.id.concat("-").concat(positionCounter.nextCount.toString())
+  );
+  if (!position) {
+    log.warning("No liquidation position found for account {} on market {}", [
+      accountId,
+      marketId,
+    ]);
+    return "";
+  }
+
+  return position.id;
+}
+
+// A series of side effects on position added
+// They include:
+// * Create a new position when needed or reuse the exisitng position
+// * Update position related data in protocol, market, account
+// * Take position snapshot
+function addPosition(
+  protocol: LendingProtocol,
+  market: Market,
+  account: Account,
+  newBalance: BigInt,
+  eventType: string,
+  side: string,
+  event: ethereum.Event
+): Position {
+  const counterID = account.id
+    .concat("-")
+    .concat(market.id)
+    .concat("-")
+    .concat(side);
+  let positionCounter = PositionCounter.load(counterID);
+  if (!positionCounter) {
+    positionCounter = new PositionCounter(counterID);
+    positionCounter.nextCount = 0;
+    positionCounter.save();
+  }
+  const positionID = positionCounter.id
+    .concat("-")
+    .concat(positionCounter.nextCount.toString());
+
+  let position = Position.load(positionID);
+  const openPosition = position == null;
+  if (openPosition) {
+    position = new Position(positionID);
+    position.account = account.id;
+    position.market = market.id;
+    position.hashOpened = event.transaction.hash.toHexString();
+    position.blockNumberOpened = event.block.number;
+    position.timestampOpened = event.block.timestamp;
+    position.side = side;
+    if (side == PositionSide.LENDER) {
+      position.isCollateral = market.canUseAsCollateral;
+    }
+    position.balance = BIGINT_ZERO;
+    position.depositCount = 0;
+    position.withdrawCount = 0;
+    position.borrowCount = 0;
+    position.repayCount = 0;
+    position.liquidationCount = 0;
+    position.save();
+  }
+  position = position!;
+  position.balance = newBalance;
+  if (eventType == EventType.DEPOSIT) {
+    position.depositCount += 1;
+  } else if (eventType == EventType.BORROW) {
+    position.borrowCount += 1;
+  }
+  position.save();
+
+  if (openPosition) {
```

Comment:
We’ve had this conversation before :D 

Try not to have the same condition evaluated in two different places. Centralize decision making.

If you think the branch of an `if` is too long because of too many things, then break it down into functions.

But evaluating the same condition multiple times can lead to nasty bugs. If at some point you need to fix something on the flow determined by that condition, it is possible that you need to modify both branches. If they are close to each other, it might be easy to catch, but the further away they are the more unlikely it will be. Not only within the same function or class, but eventually they could be in separate modules or separate services, being super hard to catch. Does this make sense?

Try to avoid it as much as possible :P

Reply:
Yeahhh,  I copied this logic from compound forks. I believe the one that fixes that issue is in the new lending schema

Reply:
Fixed